File: benchmark/mandelbrot.ts
const BAILOUT = 16.0;
const MAX_ITERATIONS = 1000;

function mandelbrot(x: number, y: number): number {
  const cr = y - 0.5;
  const ci = x;
  let zi = 0.0;
  let zr = 0.0;
  let i = 0;

  while (true) {
    i++;
    const temp = zr * zi;
    const zr2 = zr * zr;
    const zi2 = zi * zi;
    zr = zr2 - zi2 + cr;
    zi = temp + temp + ci;

    if (zi2 + zr2 > BAILOUT) {
      return i;
    }

    if (i > MAX_ITERATIONS) {
      return 0;
    }
  }
}

function main(): void {
  const start = performance.now();

  for (let y = -39; y <= 39; y++) {
    let line = "";
    for (let x = -39; x <= 39; x++) {
      const result = mandelbrot(x / 40.0, y / 40.0);
      line += result === 0 ? "*" : " ";
    }
    console.log(line);
  }

  const elapsed = (performance.now() - start) / 1000;
  console.log(`TypeScript Elapsed: ${elapsed.toFixed(2)}`);
}

main();

File: benchmark/mandelbrot.kt
const val BAILOUT = 16
const val MAX_ITERATIONS = 1000

fun mandelbrot(x: Float, y: Float): Int {
    val cr = y - 0.5f
    val ci = x
    var zi = 0.0f
    var zr = 0.0f
    var i = 0

    while (true) {
        i++
        val temp = zr * zi
        val zr2 = zr * zr
        val zi2 = zi * zi
        zr = zr2 - zi2 + cr
        zi = temp + temp + ci
        if (zi2 + zr2 > BAILOUT)
            return i
        if (i > MAX_ITERATIONS)
            return 0
    }
}

fun main() {
    val startTime = System.nanoTime()

    for (y in -39..39) {
        println()
        for (x in -39..39) {
            val result = mandelbrot(x / 40.0f, y / 40.0f)
            print(if (result == 0) "*" else " ")
        }
    }

    val endTime = System.nanoTime()
    val elapsed = (endTime - startTime) / 1_000_000_000.0
    println("\nKotlin Elapsed: %.2f".format(elapsed))
}
